akka-actor-typed/src/main/scala/akka/actor/typed/delivery/WorkPullingProducerController.scala

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (C) 2019-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.delivery
+
+import java.util.Optional
+
+import scala.reflect.ClassTag
+import scala.compat.java8.OptionConverters._
+
+import akka.Done
+import akka.actor.typed.ActorRef
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.Behavior
+import akka.actor.typed.delivery.internal.WorkPullingProducerControllerImpl
+import akka.actor.typed.receptionist.ServiceKey
+import akka.actor.typed.scaladsl.Behaviors
+import akka.annotation.ApiMayChange
+import com.typesafe.config.Config
+
+/**
+ * Work pulling is a pattern where several worker actors pull tasks in their own pace from
+ * a shared work manager instead of that the manager pushes work to the workers blindly
+ * without knowing their individual capacity and current availability.
+ *
+ * The `WorkPullingProducerController` can be used together with [[ConsumerController]] to
+ * implement the work pulling pattern.
+ *
+ * One important property is that the order of the messages should not matter, because each
+ * message is routed randomly to one of the workers with demand. In other words, two subsequent
+ * messages may be routed to two different workers and processed independent of each other.
+ *
+ * A worker actor (consumer) and its `ConsumerController` is dynamically registered to the
+ * `WorkPullingProducerController` via a [[ServiceKey]]. It will register itself to the
+ * * [[akka.actor.typed.receptionist.Receptionist]], and the `WorkPullingProducerController`
+ * subscribes to the same key to find active workers. In this way workers can be dynamically
+ * added or removed from any node in the cluster.
+ *
+ * The work manager (producer) actor will start the flow by sending a [[WorkPullingProducerController.Start]]
+ * message to the `WorkPullingProducerController`. The `ActorRef` in the `Start` message is
+ * typically constructed as a message adapter to map the [[WorkPullingProducerController.RequestNext]]
+ * to the protocol of the producer actor.
+ *
+ * The `WorkPullingProducerController` sends `RequestNext` to the producer, which is then allowed
+ * to send one message to the `WorkPullingProducerController` via the `sendNextTo` in the `RequestNext`.
+ * Thereafter the producer will receive a new `RequestNext` when it's allowed to send one more message.
+ * It will send a new `RequestNext` when there are demand from any worker.
+ * It's possible that all workers with demand are deregistered after the `RequestNext` is sent and before
+ * the actual messages is sent to the `WorkPullingProducerController`. In that case the message is
+ * buffered and will be delivered when a new worker is registered or when there is new demand.
+ *
+ * The producer and `WorkPullingProducerController` actors are supposed to be local so that these messages are
```

Comment:
And this is enforced by there being no serializer? Do you think it is worth having a way to exclude types from ever been serialized with the Java serializer?

Reply:
added enforcement check by looking at ref.path.address.hasGlobalScope